url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/sts/nusmv/NuSMVConsole.java

```diff
@@ -0,0 +1,289 @@
+//		currentState = initialNode.getState();; 		
+		int runNr = 1;
+		String fileContents = "";
+		
+		while(!strTotal.endsWith(cmd_prompt) && runNr <= NuSMVConsole.simulationSteps) {
+			try {
+				Thread.sleep(100);
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+			
+			boolean finished = false;
+			boolean noUpdate = true;
+			
+			while (noUpdate && !finished) {
+				fileContents = IOUtils.readFileAsString(outFile);
+				if (strTotal.endsWith(cmd_prompt) || 
+						fileContents.contains("Trace Description")) {
+					finished = true;
+					break;
+				}
+				
+				String[] runs = fileContents.split("AVAILABLE STATES");
+				if (runs.length - 1 == runNr) {
+					noUpdate = false;
+					break;
+				} 
+				// nothing new was added, wait a while
+				try {
+					Thread.sleep(100);
+//					logger.debug("Nothing new was added, waiting..");
+				} catch (InterruptedException e) {
+					e.printStackTrace();
+				}
+			}
+			
+			if (!finished) {
+				pickNextState(fileContents);
+			}
+			runNr++;
+		}
+		
+		synchronized(this) {
+			return printConsole();
+		}
+	}
+	
+	
+	private void pickNextState(String fileContents) throws IOException {
+		
+		List<NusmvState> states = 
+			new ArrayList<NusmvState>();
+		NusmvState currentState = (NusmvState) this.currentNode.getState();
+		
+		try {
+			states = NuSMVOutputHandler.addTransitionsFromState(
+					fileContents, currentState, sts);
+		} catch (IllegalArgumentException e) {
+			// read the file too soon, one action was not complete 
+			try {
+				logger.info("Read file to soon, waiting..");
+				Thread.sleep(200);
+			} catch (InterruptedException ie) {
+				ie.printStackTrace();
+			}
+			fileContents = IOUtils.readFileAsString(outFile);
+			states = NuSMVOutputHandler.addTransitionsFromState(
+					fileContents, currentState, sts);
+		}
+		
+		if (!states.isEmpty()) {
+			List<Node> nodes = currentNode.addChildren(states);
+			int index = 0;
+			if (nodes.size() > 1) {		
+				for (int i = 0; i < nodes.size(); i++) {
+					if (!nodes.get(i).subtreeIsVisited()) {
+						index = i;
+						break;
+					}
+				}
+			}
+			if (nodes.size() >= 1) {
+				currentNode = nodes.get(index);
+//				currentState = states.get(index);
+				currentNode.visit();
+				execCommand(String.valueOf(index));
+			}				
+		}
+	}
+	
+	
+	void loadConsole() throws IOException {
+		File file = new File(nusmvFile);
+		if (!file.exists()) {
+			// TODO Throw an exception
+			logger.error("Cannot find file " + nusmvFile);
+		}
+		
+		String[] cmdarray = {nusmvPath + "NuSMV.exe", "-int", nusmvFile};
+		ProcessBuilder builder = new ProcessBuilder(cmdarray);
+		builder.redirectErrorStream(true);		
+		proc = new ProcessBuilder(cmdarray).start();
+
+		InputStream inputStream = proc.getInputStream();
+		OutputStream outputStream = proc.getOutputStream();
+		brInput = new BufferedReader(new InputStreamReader(inputStream));
+		bwOutput = new BufferedWriter(new OutputStreamWriter(outputStream));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *bwOutput*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).